### DISCUSSION TITLE
How can I configure a neovim keybind launching a command not to queue up keyboard input?

### DISCUSSION THREAD

**Discussion by unphased** [neovim/neovim#29905] [Q&A] [answered]
Here is a test case: 

- Run `:lua vim.fn.system('sleep 5');`
- Before 5 seconds is up, type a bunch of keys that will be input into neovim in normal mode. For example spam `llllll` to move the cursor to the right. 
- Observe: During the 5 seconds no input is sent to neovim. The call to `vim.fn.system` is indeed a blocking call.
- Observe: **After 5 seconds elapses**, all keyboard input is inserted into neovim normal mode. (e.g. cursor moves to the right, all at once, after the system call finishes.

EDIT: I looked in documentation and refactored my lua code from using `vim.fn.system` which would use the VimL `system` into using `vim.system` (including a call to `:wait()` after) which is preferred, I still get the same behavior. 

**_DESIRED BEHAVIOR_**: a way to call vim.fn.system or maybe a mode change (possibly as an instruction to the terminal to ignore input???) to make it so that input provided to the terminal is ignored. 

Actual use case: 

I call `tmux select-window -L` (or `-D`, or `-U`, or `-R`) from nvim to implement contextual window navigation: If i am trying to move to the window toward the right with my binds and neovim's focused window is already the right-most window, then I inform tmux that I want it to focus the pane to the right. 

There is a race condition that exists. My above functionality is bound to `<C-l>`, and if I type <kbd>Ctrl</kbd>+<kbd>L</kbd>,<kbd>q</kbd> too quickly, with the intent of 

1. going to the window to the right (in this case a tmux pane to the right of the currently focused neovim)
2. sending q to the program running in that terminal

The race condition is that my typed <kbd>q</kbd> is sent to tmux and neovim before neovim finishes executing `tmux select-pane -R`. 

1. This causes neovim to consume the `q` input of course (since tmux hasnt finished resolving the command to change the focused pane) 
2. `q` becomes sent into nvim insert mode (now it's ready to accept a character for a recording register)
3. tmux finally brings the focus to the intended terminal pane

I am aware that architecturally speaking there is no way in this situation to allow `q` to actually land in the intended tmux pane, that being a limitation of terminals and tmux at a base level, but I'd like to at least be able to suppress the `q` from landing in nvim normal mode. 

Is this possible?

**Comment by unphased** [comment]
Claude 3.5 Sonnet gave me a few suggestions, two of which involve manually intervening with input by providing a callback to `vim.on_key()` which is temporarily configured to consume and ignore keyboard input. I would then be able to add a command after my `tmux select-pane` call that disables it. I don't see why this wouldn't work, but it's somewhat inelegant. 

Upon further research along these lines, it occurs to me that assembling a temporary key consuming mode here would allow me to record the keys as necessary and **replay them back to tmux via `tmux send-keys`**. I will report back if I attempt this.

**Reply by knarrff** [reply]
> Upon further research along these lines, it occurs to me that assembling a temporary key consuming mode here would allow me to record the keys as necessary and **replay them back to tmux via `tmux send-keys`**. I will report back if I attempt this.

I would think this could you in even more trouble, as you wouldn't exactly know when to send them and you could have already typed other things into the tmux session before the stuff from before is sent.

**Reply by unphased** [reply]
Yes, it appears like it should give rise to MULTIPLE further race conditions and I should expect the sequence of inputs landing in the target tmux pane to become interlaced. 

That being said... per original use case if the amount of time involved in these race conditions is small such that i can only really get in one character edgewise like my example `q`, that would potentially still guarantee my original intent.

**Comment by knarrff** [comment, marked as answer]
I am thinking: maybe it is possible to do this the other way around: define the shortcut in tmux. Let tmux send a command to a running vim session when one is detected, possibly looking for the return value (whether you are already at the right-most pane) and then decide right there. My thinking is: at this point tmux should be blocking input anyway, so the input would automatically go to whereever tmux switched (or not).

I haven't done this myself, but ```:h clientserver``` sounds promising.

**Reply by unphased** [reply]
Ah I think you're onto something here. 

This definitely has legs. I've been through many iterations of deep tmux integrations. Back in the day I used to have a shell script that would implement tmux based contextual actions, which I ported to C++ at one point to get snappier behavior and tighter windows on the race conditions. Then in the tmux 3.2/3.3 timeframe it gained built in logical operators in expansions that I could use to directly implement the logic I needed in complex tmux binds themselves, and that has indeed eliminated all of *those* race conditions. 

As far as I know of the tmux key bind execution model this should be possible. Yes, it hinges on whether I can ask neovim for this state in a blocking way within the tmux keybind.

**Reply by unphased** [reply]
Another benefit to putting this behavior in on the tmux side is to avoid the ambiguity involved across tmux instances and sessions whenever we make a call to `tmux`. Some of my config does not take this into account so In some rare cases I'll have integrations send commands to some other tmux instance i had running, leading to some temporary confusion.

**Reply by unphased** [reply]
re: `clientserver` I already use `nvr` for some hyperlink streamlining neovim integration purposes so I reckon that would be the ticket for me. Likely also possible to go without it to have even fewer dependencies but the command would probably be nasty.
